# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/ViewportUtils.cpp
# Commit: b76b8d0b94ca
# Full Hash: b76b8d0b94cab1edac764542d3d36604d629746c
# Author: Botond Ballo <botond@mozilla.com>
# Date: 2020-05-06 09:37:16
# Regressor Bug: 1556556
# File Overlap Count: 1
# Description:
#   Bug 1556556 - Add some VisualToLayout() and LayoutToVisual() wrappers to ViewportUtils. r=kats
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D68724
# ==============================================================================

diff -r ff865d99f67e -r b76b8d0b94ca layout/base/ViewportUtils.cpp
--- a/layout/base/ViewportUtils.cpp	Tue May 05 19:37:26 2020 +0000
+++ b/layout/base/ViewportUtils.cpp	Tue May 05 19:37:26 2020 +0000
@@ -5,6 +5,7 @@
 #include "mozilla/PresShell.h"
 #include "mozilla/ViewportUtils.h"
 #include "mozilla/layers/APZCCallbackHelper.h"
+#include "mozilla/layers/InputAPZContext.h"
 #include "mozilla/layers/ScrollableLayerGuid.h"
 #include "nsIContent.h"
 #include "nsLayoutUtils.h"
@@ -12,6 +13,7 @@
 namespace mozilla {
 
 using layers::APZCCallbackHelper;
+using layers::InputAPZContext;
 using layers::ScrollableLayerGuid;
 
 CSSToCSSMatrix4x4 ViewportUtils::GetVisualToLayoutTransform(
@@ -47,4 +49,49 @@
       .PostTranslate(transform.x, transform.y, 0);
 }
 
+CSSToCSSMatrix4x4 GetVisualToLayoutTransform(PresShell* aContext) {
+  ScrollableLayerGuid::ViewID targetScrollId =
+      InputAPZContext::GetTargetLayerGuid().mScrollId;
+  if (targetScrollId == ScrollableLayerGuid::NULL_SCROLL_ID) {
+    if (nsIFrame* rootScrollFrame = aContext->GetRootScrollFrame()) {
+      targetScrollId =
+          nsLayoutUtils::FindOrCreateIDFor(rootScrollFrame->GetContent());
+    }
+  }
+  return ViewportUtils::GetVisualToLayoutTransform(targetScrollId);
+}
+
+nsPoint ViewportUtils::VisualToLayout(const nsPoint& aPt, PresShell* aContext) {
+  auto visualToLayout = mozilla::GetVisualToLayoutTransform(aContext);
+  CSSPoint cssPt = CSSPoint::FromAppUnits(aPt);
+  cssPt = visualToLayout.TransformPoint(cssPt);
+  return CSSPoint::ToAppUnits(cssPt);
+}
+
+nsRect ViewportUtils::VisualToLayout(const nsRect& aRect, PresShell* aContext) {
+  auto visualToLayout = mozilla::GetVisualToLayoutTransform(aContext);
+  CSSRect cssRect = CSSRect::FromAppUnits(aRect);
+  cssRect = visualToLayout.TransformBounds(cssRect);
+  nsRect result = CSSRect::ToAppUnits(cssRect);
+
+  // In hit testing codepaths, the input rect often has dimensions of one app
+  // units. If we are zoomed in enough, the rounded size of the output rect
+  // can be zero app units, which will fail to Intersect() with anything, and
+  // therefore cause hit testing to fail. To avoid this, we expand the output
+  // rect to one app units.
+  if (!aRect.IsEmpty() && result.IsEmpty()) {
+    result.width = 1;
+    result.height = 1;
+  }
+
+  return result;
+}
+
+nsPoint ViewportUtils::LayoutToVisual(const nsPoint& aPt, PresShell* aContext) {
+  auto visualToLayout = mozilla::GetVisualToLayoutTransform(aContext);
+  CSSPoint cssPt = CSSPoint::FromAppUnits(aPt);
+  auto transformed = visualToLayout.Inverse().TransformPoint(cssPt);
+  return CSSPoint::ToAppUnits(transformed);
+}
+
 }  // namespace mozilla
\ No newline at end of file